### PR TITLE
fix: allow to delete all ACLs for principal

### PR DESCRIPTION
**Pull request by wtrocki**
Creating for early feedback for https://github.com/redhat-developer/app-services-cli/issues/1430

Validation for delete command were redundant preventing users from deleting multiple acls. 
For example it would not be possible to remove all ACLs for Service account.

After this change it is possible:

```
✔️  Deleted 5 ACLs from Kafka instance "wtrocki"
The following ACL rules were deleted:

  PRINCIPAL (5)    PERMISSION   OPERATION   DESCRIPTION               
 ---------------- ------------ ----------- -------------------------- 
  test             allow        write       transactional-id is "*"   
  test             allow        describe    topic starts with "test"  
  test             allow        write       topic starts with "test"  
  test             allow        create      topic starts with "test"  
  test             allow        describe    transactional-id is "*"   
  ```

However we need to still spend some time cleaning up messages/validation and arguments to API.
Posted for early feedback

## Verification
```
[wtrocki@graphapi app-services-cli (delete-acl)]$ rhoas kafka acl delete --pattern-type=any --service-account=test
? All ACLs matching the criteria provided will be deleted from the Kafka instance "wtrocki". Are you sure you want to proceed? Yes


✔️  Deleted 5 ACLs from Kafka instance "wtrocki"
The following ACL rules were deleted:

  PRINCIPAL (5)    PERMISSION   OPERATION   DESCRIPTION               
 ---------------- ------------ ----------- -------------------------- 
  test             allow        describe    topic starts with "test"  
  test             allow        write       transactional-id is "*"   
  test             allow        describe    transactional-id is "*"   
  test             allow        write       topic starts with "test"  
  test             allow        create      topic starts with "test" 
  ```

Delete all write operations for sa
```
rhoas kafka acl delete --service-account=test1 --operation=write -v --pattern-type=any
```

other variations

### CODE REVIEW

**Comment by rkpattnaik780** [comment]
Deleting all acls for a principal.
Passing other arguments needs to be handled for deleting specific ACL. Ignore if it is on todo.

**Comment by wtrocki** [comment]
Not sure if we should allow all accounts as well?

```

[wtrocki@graphapi app-services-cli (delete-acl ✗)]$ rhoas kafka acl delete --all-accounts         
? All ACLs matching the criteria provided will be deleted from the Kafka instance "wtrocki". Are you sure you want to proceed? Yes


✔️  Deleted 4 ACLs from Kafka instance "wtrocki"
The following ACL rules were deleted:

  PRINCIPAL (4)    PERMISSION   OPERATION          DESCRIPTION     
 ---------------- ------------ ------------------ ---------------- 
  All Accounts     allow        describe-configs   topic is "*"    
  All Accounts     allow        describe           group is "*"    
  All Accounts     allow        describe           cluster is "*"  
  All Accounts     allow        describe           topic is "*"  
  
  ```

**Comment by wtrocki** [comment]
Seeing some strange issue/inconsitencies:

```
[wtrocki@graphapi app-services-cli (delete-acl)]$ rhoas kafka acl list                    

  PRINCIPAL (5)    PERMISSION   OPERATION   DESCRIPTION               
 ---------------- ------------ ----------- -------------------------- 
  wtrocki          allow        create      topic starts with "test"  
  wtrocki          allow        describe    topic starts with "test"  
  wtrocki          allow        write       topic starts with "test"  
  wtrocki          allow        describe    transactional-id is "*"   
  wtrocki          allow        write       transactional-id is "*"   
[wtrocki@graphapi app-services-cli (delete-acl)]$ rhoas kafka acl delete --service-account=wtrocki
? All ACLs matching the criteria provided will be deleted from the Kafka instance "wtrocki". Are you sure you want to proceed? Yes


✔️  Deleted 2 ACLs from Kafka instance "wtrocki"
The following ACL rules were deleted:

  PRINCIPAL   PERMISSION   OPERATION   DESCRIPTION              
 ----------- ------------ ----------- ------------------------- 
  wtrocki     allow        describe    transactional-id is "*"  
  wtrocki     allow        write       transactional-id is "*"  
  
```

**Comment by wtrocki** [comment]
Ok. So problem is with Pattern Type. because we assume Literal pattern type is provided by default there is no easy way for us to fix that without breaking CLI. I will need to add explicit pattern-type=none and deprecated --prefix to get around this

**Comment by wtrocki** [comment]
@craicoverflow @rkpattnaik780 I have managed to workaround problem by introducing new flag called pattern type and deprecated prefix. 

Not sure if we should be deprecating prefix. I'm testing all gazillion test use cases that this change introduced

**Comment by craicoverflow** [comment]
Pattern type is fine to use. I used "prefix" because this is what confluent CLI does.

**Comment by rkpattnaik780** [comment]
> Not sure if we should allow all accounts as well?

UI allows deleting permissions for all accounts. We should keep it too.

File: pkg/cmd/kafka/acl/delete/delete.go [hunk 1]
```diff
@@ -226,11 +256,20 @@ func validateAndSetOpts(opts *aclcmdutil.CrudOptions) error {
 		return opts.Localizer.MustLocalizeError("kafka.acl.common.error.noPrincipalsSelected")
 	}
 
-	opts.PatternType = aclcmdutil.PatternTypeLITERAL
+	// Backwards compatibility:
```
Comment (rkpattnaik780) [inline pkg/cmd/kafka/acl/delete/delete.go:259]: This should be moved after the switch in 264.
Current behavior keeps pattern type as literal if we pass `--prefix` flag.

```
./rhoas kafka acl delete --user enda --topic "rhoas-" --prefix -v
```
```
/rest/acls?patternType=LITERAL&permission=ANY&principal=User%3Aenda&resourceName=rhoas-&resourceType=TOPIC
```

File: pkg/core/localize/locales/en/cmd/acl.en.toml [hunk 2]
```diff
@@ -85,6 +85,9 @@ one = 'Set the resource type to cluster'
 [kafka.acl.common.flag.prefix.description]
 one = 'Determine if the resource should be exact match or prefix'
 
+[kafka.acl.common.flag.patterntypes.description]
+one = 'Determine if the resource should be exact match, prefix or any {{.Types}}'
```
Comment (rkpattnaik780) [inline pkg/core/localize/locales/en/cmd/acl.en.toml:89]: The flag description looks odd

```
Determine if the resource should be exact match, prefix or any [any literal prefix] (default "literal")
```

Should we keep options in description only?

**Review by rkpattnaik780 — CHANGES_REQUESTED** [review]

File: pkg/cmd/kafka/acl/delete/delete.go (hunk 1 above)
Comment (wtrocki) [reply pkg/cmd/kafka/acl/delete/delete.go:259]: Oh.. I did refactor if into switch and introduced that bug. Really hate that golanglint forces us to use switch statements even when they do not make sense

File: pkg/core/localize/locales/en/cmd/acl.en.toml (hunk 2 above)
Comment (wtrocki) [reply pkg/core/localize/locales/en/cmd/acl.en.toml:89]: It is good to give list of values in canonical form. Will remove them from description

**Comment by craicoverflow** [comment]
Screenshot 2022-02-17 at 09 12 40

What is this submodule for?

**Comment by wtrocki** [comment]
> What is this submodule for?

That was different POC. Removing it

**Review by craicoverflow — APPROVED** [review]
Verified. I left one suggestion regarding printing deprecation message from within the command output aswell as in the description.

**Comment by wtrocki** [comment]
> You could always look to remove that rule from the lint config file? It is not a rule coming from the Go stdlib or anything.

Rule itself is good, but it does have lots of false positives - where doing switch is not possible. 
I will need to start disabling linting for this instead.

File: pkg/core/localize/locales/en/cmd/acl.en.toml [hunk 3]
```diff
@@ -291,6 +294,9 @@ $ rhoas kafka acl delete --operation write --permission allow --topic all --user
 # Delete an ACL for a service account
 $ rhoas kafka acl delete --operation all --permission any --topic "rhoas" --prefix --service-account "srvc-acct-11924479-43fe-42b4-9676-cf0c9aca81"
 
+# Delete all ACLs for a service account
+$ rhoas kafka acl delete --service-account "srvc-acct-11924479-43fe-42b4-9676-cf0c9aca81 --pattern-type=all"
```
[the remark this answers is not in the document]
Comment (wtrocki) [reply pkg/core/localize/locales/en/cmd/acl.en.toml:298]: Yes. While it is counter intuitive we need backwards compatibility.

So:

- nothing provided - exact match
- all - take all
- prefix (use prefix)

etc.

**Review by rkpattnaik780 — APPROVED** [review]

File: pkg/cmd/kafka/acl/delete/delete.go [hunk 4]
```diff
@@ -92,6 +91,19 @@ func NewDeleteCommand(f *factory.Factory) *cobra.Command {
...
+		&patternTypeFlag,
+		"pattern-type",
+		aclcmdutil.PatternTypeLITERAL,
+		opts.Localizer.MustLocalize("kafka.acl.common.flag.patterntypes.description",
+			localize.NewEntry("Types", aclcmdutil.PatternTypes)),
+	)
+
+	_ = cmd.RegisterFlagCompletionFunc("pattern-type", func(cmd *cobra.Command, _ []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+		return aclcmdutil.PatternTypes, cobra.ShellCompDirectiveNoSpace
+	})
```
[the remark this answers is not in the document]
Comment (craicoverflow) [reply pkg/cmd/kafka/acl/delete/delete.go:106]: The package already exists though, so don't see how (there is no extra code involved).

File: pkg/core/cmdutil/flagutil/deprecation.go [hunk 5]
```diff
@@ -0,0 +1,6 @@
+package flagutil
+
+// DeprecateFlag provides a way to deprecate a flag by appending standard prefixes to the flag description.
+func DeprecateFlag(flagDescription string) string {
+	return "DEPRECATED: " + flagDescription
+}
```
[the remark this answers is not in the document]
Comment (wtrocki) [reply pkg/core/cmdutil/flagutil/deprecation.go:6]: Fully agree.  I will remove that method as it is not used right now and it is not properly implemented.

**Pull request merged**